### DISCUSSION TITLE
`Astro.context`

### DISCUSSION THREAD

**Discussion by pilcrowonpaper** [withastro/roadmap#370] [Proposal]
This is based on #230, but I felt there was enough differences for it to at least warrant a new discussion. Inspiration is mainly from SvelteKit's `getSession()` (now removed) and hooks.

## Issue

If you want to run the same logic on each request, you have to handle that individually in each page. One example this happens is authentication, where you check for the session cookie for each request:

```ts
// src/pages/index.astro
const { auth_session } = cookie.parse(Astro.request.headers.get("cookie");
const session = await validateSession(auth_session);
```

Ideally, you'll add all that inside a single function that runs before each request.

## Proposal

The proposal is to introduce `app.ts` inside `src` directory. This file can export `setContext()`, which will be called before rendering the page. This means it will be called **only during the build-step for SSG, or on the build-step and on each request for SSR**. This means it will work and have similar functionalities regardless of the adapter used. The returned object will be applied to `Astro.context`, which will available in all pages.

```ts
// src/app.ts
export const setContext = () => {
   // ...
};
```

The type for `setContext()` is the following:

```ts
type setContext: (request: Request, response: Response) => Promise<Record<any, any>>;
```

where `Request` is the standard `Astro.request` and `Response` is the standard `Astro.response`. I've included `Response` as well since the values of the returned context object can be a function and you may want to include a function that modifies the response object (eg. set cookie). I'm sure Astro can throw an error when using `Astro.response` during SSG.

*There may be a better name for `app.ts`*

### Example

```ts
// src/app.ts
export const setContext = async (request, response) => {
   const { auth_session } = cookie.parse(request.headers.get("cookie");
   const session = await validateSession(auth_session);
   return {
      session,
      getUser: () => getUser(session); // can be a function
      setRandomCookie: () => response.headers.set('Set-Cookie', 'a=b; Path=/;');
   };
};
```

```ts
// src/pages/index.astro
const session = Astro.context.session;
const user = await Astro.context.getUser();
Astro.context.setRandomCookie() // modify response
```

### `setContext()` parameters

I'm not sure if `request, response` should be just the `Astro` global object. However, I went against `Astro` since it provides page-specific things such as `Astro.self`. 

### `locals`

Express and other frameworks provide `locals` (like `app.locals`), which allows data to be sent between middleware and the request handler. One advantage this proposal has over using `app.locals` is type safety. While it's possible to pre-declare the type of locals using namespaces like so (example from SvelteKit):

```ts
declare global {
    namespace App {
        interface Locals {
            session: Session;
        }
    }
}
```

there's no guarantee that `locals.session` will exist since there's no way to enforce that `locals.session` is set inside middleware.

### Type safety

Maybe there are better ways to handle this but here's what I came up with:

#### Option 1

Similar to the example above, we can let the user set the types for `Astro.context` using namespaces:

```ts
// src/env.d.ts
namespace App {
    export interface Context {
        session: Session;
    }
}
```

Astro can then export the following type:

```ts
// src/app.ts
export type SetContext: (request: Request) => Promise<App.Context>
```

Which can be imported inside `src/app.ts`:

```ts
// src/app.ts
export const setContext: SetContext = () => {
   return {
      session
   }:
};
```

#### Option 2

Similar to option 1, but since the existence of `src/app.ts` file is guaranteed, we can get the type of `setContext()` and import it into the namespace so the return type is accessible across the app.

```ts
// src/app.ts
export const setContext = () => {
   return {
      session
   }:
};

export type SetContext = typeof setContext
```

```ts
// src/env.d.ts
namespace App {
    export type SetContext = import("./app.ts").SetContext
}
```

```ts
type Astro = {
   context: Awaited<ReturnType<App.SetContext>>
}
```

**Comment by aFuzzyBear** [comment]
Hi @pilcrowOnPaper this is really nice, there is indeed alot of interest and thought going into a potential `Astro.context` API, and this has helped add some great ideas to it. So thank you for that. My personal thoughts are that the need for an `Astro.context` layer is sound and valid. Implementation seems to be a sticking point in some areas. your idea of having an `App.ts` in the base of the route is interesting, however I feel that this stuff might be better handled by an `Astro.middleware.context`. 
Again really interesting, 💚 👊

**Reply by pilcrowonpaper** [reply]
A full on middleware implementation would work too but I'd like a potential `Astro.locals` to have some type safety, which can only be done if the value is set via a function.

**Comment by florian-lefebvre** [comment]
Closing because this is outdated
